sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -419,8 +422,13 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       }
 
       if (mode == SaveMode.Overwrite && tableExists) {
-        JdbcUtils.dropTable(conn, table)
-        tableExists = false
+        if (extraOptions.getOrElse("truncate", "false").toBoolean &&
+            JdbcUtils.isCascadingTruncateTable(url) == Some(false)) {
+          JdbcUtils.truncateTable(conn, table)
```

Comment:
- Drop, Create and Insert: Create and Insert could fail, but we still drop the table. 
- Truncate and Insert: Insert could fail, but we always truncate the table. 

I think it is OK to raise an exception here, but check whether the exception message is meaningful or not.